lib/awful/popup.lua

```diff
@@ -0,0 +1,467 @@
+            self,
+            context,
+            self._private.widget,
+            self._wb._private.maximum_width  or 9999,
+            self._wb._private.maximum_height or 9999
+        )
+        timer.delayed_call(function()
+            apply_size(self._wb, w, h, true)
+        end)
+        return { wibox.widget.base.place_widget_at(self._private.widget, 0, 0, width, height) }
+    end
+end
+
+-- Set the widget that is drawn on top of the background
+function main_widget:set_widget(widget)
+    if widget then
+        wibox.widget.base.check_widget(widget)
+    end
+    self._private.widget = widget
+    self:emit_signal("widget::layout_changed")
+end
+
+function main_widget:get_widget()
+    return self._private.widget
+end
+
+function main_widget:get_children_by_id(name)
+    return self._wb:get_children_by_id(name)
+end
+
+local popup = {}
+
+--- Set the preferred popup position relative to its parent.
+--
+-- This allows, for example, to have a submenu that goes on the right of the
+-- parent menu. If there is no space on the right, it tries on the left and so
+-- on.
+--
+-- Valid directions are:
+--
+-- * left
+-- * right
+-- * top
+-- * bottom
+--
+-- The basic use case for this method is to give it a parent wibox:
+--
+-- @DOC_awful_popup_position1_EXAMPLE@
+--
+-- As demonstrated by this second example, it is also possible to use a widget
+-- as a parent object:
+--
+-- @DOC_awful_popup_position2_EXAMPLE@
+--
+-- @property preferred_positions
+-- @tparam table|string preferred_positions A position name or an ordered
+--  table of positions
+-- @see awful.placement.next_to
+-- @see awful.popup.preferred_anchors
+
+function popup:set_preferred_positions(pref_pos)
+    self._private.preferred_directions = pref_pos
+    set_position(self)
+end
+
+--- Set the preferred popup anchors relative to the parent.
+--
+-- The possible values are:
+--
+-- * front
+-- * middle
+-- * back
+--
+-- For details information, see the `awful.placement.next_to` documentation.
+--
+-- In this example, it is possible to see the effect of having a fallback
+-- preferred anchors when the popup would otherwise not fit:
+--
+-- @DOC_awful_popup_anchors_EXAMPLE@
+--
+-- @property preferred_anchors
+-- @tparam table|string preferred_anchors Either a single anchor name or a table
+--  ordered by priority.
+-- @see awful.placement.next_to
+-- @see awful.popup.preferred_positions
+
+function popup:set_preferred_anchors(pref_anchors)
+    self._private.preferred_anchors = pref_anchors
+    set_position(self)
+end
+
+--- The current position relative to the parent object.
+--
+-- If there is a parent object (widget, wibox, wibar, client or the mouse), then
+-- this property returns the current position. This is determined using
+-- `preferred_positions`. It is usually the preferred position, but when there
+-- isn't enough space, it can also be one of the fallback.
+--
+-- @property current_position
+-- @param string Either "left", "right", "top" or "bottom"
```

Comment:
This looks different than all the others. Should it be this?
`-- @tparam string current_position Either "left", "right", "top" or "bottom"`

Reply:
The property above, L188, is a `@tparam`:
```
-- @property preferred_anchors
-- @tparam table|string preferred_anchors Either a single anchor name or a table
--  ordered by priority.
```
The property below, L221, is a `@tparam` and exactly the same as `current_position` only with different acceptable strings:
```
-- @property current_anchor
-- @tparam string current_anchor Either "front", "middle", "back"
```
So why isn't this one `@tparam` like the others?

Reply:
Oh, that, ok. I was thinking about the property name or the accepted strings.

Fixed. Thanks for the review.